### ISSUE TITLE
Visual bell persists past end of animation

### ISSUE THREAD

**Issue by virtualdxs** [alacritty/alacritty#3083]
I'm still seeing what I thought was #2914 on 0.4.0. I cause a bell, it mostly finishes the animation, but there's still an overlay of red (bell color) until I type another key. The opaqueness varies from bell to bell.

I'm able to reproduce on the below system but not my main laptop.
### Bell configuration:

    visual_bell:
      animation: Linear
      duration: 200
      color: '0xff0000'

### System

OS: Arch Linux
Version: alacritty 0.4.0
Xorg version 1.20.6 and dwm
GPU: 

    08:00.0 VGA compatible controller: NVIDIA Corporation GK208B [GeForce GT 710] (rev a1)

**Comment by chrisduerr** [comment]
Can you reproduce this with just those four lines in your configuration file, without anything else?

DWM is also always a bit tricky, since it doesn't support a lot of stuff. Do you have a compositor running? Do you have access to a system with a different WM?

Which command are you running to cause this? And could you verify the exact Alacritty version with `alacritty --version`?

**Comment by dannydulai** [comment]
@chrisduerr - I was able to reproduce with just those 4 lines in the config file.

```
% /Applications/Alacritty.app/Contents/MacOS/alacritty  --version
alacritty 0.4.0 (b115b90)
% uname -a
Darwin thoo.local 19.2.0 Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64 x86_64
```

Running on latest Macbook Pro 16, fully updated MacOS as of today.

**Comment by chrisduerr** [comment]
@dannydulai Thanks for letting me know this can be reproduced on macOS. I've got two Arch machines and cannot reproduce on either, but I'll try on my macbook once I have access to it again.

**Comment by dannydulai** [comment]
When I'm at my shell prompt and I hit backspace, it beeps (visually). If I hit backspace firmly and cleanly, it works fine. If I tap it very lightly and quickly, I'm left with a transparent red overlay.

**Comment by chrisduerr** [comment]
But it goes away as soon as you hit backspace again, right? How consistently can you reproduce this?

**Comment by dannydulai** [comment]
the visual bell goes away as soon as I do anything... even moving the mouse (even if the mouse is not hovering over the window). Hitting backspace again just triggers the bell. Then, as usual, it finishes the animation and clears out or it doesn't.

I can reproduce it very reliably, and I can also reproduce it the same way, using the same alacritty build on another Mac Book Pro from 2016 running a much older operating system (Sierra).

I usually run in "fake fullscreen mode", but it happens in windowed mode too.

**Comment by haarp** [comment]
Same issue here. Alacritty 0.4.0 on Linux X11 with these settings:

```visual_bell:
  animation: Ease
  duration: 75```

This seems to be related to using nvidia cards. When I run X on the integrated Intel graphics of my laptop, this issue does not appear. When I run X on the nvidi card (with nvidia driver), this issue presents itself.

**Comment by chrisduerr** [comment]
> This seems to be related to using nvidia cards. When I run X on the integrated Intel graphics of my laptop, this issue does not appear. When I run X on the Nvidia GPU (with nvidia-440.44 driver), this issue presents itself.

That's an interesting point, since I do not own any Nvidia GPUs and I cannot reliably reproduce. I think I was able to reproduce once, but it was very difficult.

Can others confirm that this is an issue with Nvidia? @dannydulai / @virtualdxs ?

**Comment by dannydulai** [comment]
Latest MacBook Pro 16 has AMD.

**Comment by virtualdxs** [comment]
I can confirm that nvidia seems to be worse about this issue. My laptop, which I could not reproduce on with its integrated graphics, does let me repro when using the nvidia GPU.

**Comment by chrisduerr** [comment]
I'd be curious if the length of the visual bell matters, or if this can be reproduced with visual bells that are multiple seconds long.

I've also got a patch that should print some more information about the status of the visual bell to stdout, if someone could send me a log of that, I'd appreciate it.

<details>

```diff
diff --git a/alacritty/src/display.rs b/alacritty/src/display.rs
index 317c8758..11a3f43b 100644
--- a/alacritty/src/display.rs
+++ b/alacritty/src/display.rs
@@ -420,6 +420,7 @@ impl Display {
         }
 
         // Push visual bell after url/underline/strikeout rects
+        println!("VISUAL BELL INTENSITY: {} -> {}", visual_bell_intensity, visual_bell_intensity != 0.);
         if visual_bell_intensity != 0. {
             let visual_bell_rect = RenderRect::new(
                 0.,
diff --git a/alacritty/src/event.rs b/alacritty/src/event.rs
index bfadbaec..3c8bc89f 100644
--- a/alacritty/src/event.rs
+++ b/alacritty/src/event.rs
@@ -431,7 +431,10 @@ impl<N: Notify + OnResize> Processor<N> {
 
                 // Request immediate re-draw if visual bell animation is not finished yet
                 if !terminal.visual_bell.completed() {
+                    println!("REQUESTING REDRAW DUE TO VISUAL BELL");
                     event_queue.push(GlutinEvent::UserEvent(Event::Wakeup));
+                } else {
+                    println!("VISUAL BELL COMPLETED");
                 }
 
                 // Redraw screen
```

</details>

If someone is able to capture this in a log, please make sure it is killed from an external source, so it does not manage to write additional lines to stdout during the shutdown process, potentially clearing the visual bell again.

**Comment by haarp** [comment]
@chrisduerr 
I just tested this on my Nvidia card. When the bell chimes this is logged

```
REQUESTING REDRAW DUE TO VISUAL BELL
VISUAL BELL INTENSITY: 0.7527031076887233 -> true
REQUESTING REDRAW DUE TO VISUAL BELL
VISUAL BELL INTENSITY: 0.7570199739382184 -> true
REQUESTING REDRAW DUE TO VISUAL BELL
VISUAL BELL INTENSITY: 0.8016586685776896 -> true
REQUESTING REDRAW DUE TO VISUAL BELL
VISUAL BELL INTENSITY: 0.7344287862984428 -> true
```

and it's stuck like this. When I move the mouse cursor, or bring the window into focus, or type into it, disappears again with this:

```
REQUESTING REDRAW DUE TO VISUAL BELL
VISUAL BELL INTENSITY: 0 -> false
VISUAL BELL COMPLETED
VISUAL BELL INTENSITY: 0 -> false
```

**Comment by chrisduerr** [comment]
@haarp That's extremely interesting, seems like one of our events is getting lost.

Could you give me the same output, with `alacritty --print-events`? That should additionally log all events being received.

**Comment by haarp** [comment]
@chrisduerr 

Bell chimes:

```
[2020-02-07 15:55] [INFO] glutin event: NewEvents(WaitCancelled { start: Instant { tv_sec: 7485, tv_nsec: 858681027 }, requested_resume: None })
[2020-02-07 15:55] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(17))), event: Key(KeyboardInput { scancode: 28, state: Released, virtual_keycode: Some(Return), modifiers: (empty) }) }
[2020-02-07 15:55] [INFO] glutin event: MainEventsCleared
[2020-02-07 15:55] [INFO] glutin event: RedrawEventsCleared
[2020-02-07 15:55] [INFO] glutin event: NewEvents(WaitCancelled { start: Instant { tv_sec: 7487, tv_nsec: 2493957 }, requested_resume: None })
[2020-02-07 15:55] [INFO] glutin event: UserEvent(Title("echo [@hellbringer] (15:55:07)"))
[2020-02-07 15:55] [INFO] glutin event: UserEvent(Urgent)
[2020-02-07 15:55] [INFO] glutin event: UserEvent(Wakeup)
[2020-02-07 15:55] [INFO] glutin event: MainEventsCleared
[2020-02-07 15:55] [INFO] glutin event: RedrawEventsCleared
REQUESTING REDRAW DUE TO VISUAL BELL
VISUAL BELL INTENSITY: 0.75480112461764 -> true
[2020-02-07 15:55] [INFO] glutin event: NewEvents(WaitCancelled { start: Instant { tv_sec: 7487, tv_nsec: 484097551 }, requested_resume: None })
[2020-02-07 15:55] [INFO] glutin event: MainEventsCleared
[2020-02-07 15:55] [INFO] glutin event: RedrawEventsCleared
REQUESTING REDRAW DUE TO VISUAL BELL
VISUAL BELL INTENSITY: 0.7606717086472313 -> true
[2020-02-07 15:55] [INFO] glutin event: NewEvents(WaitCancelled { start: Instant { tv_sec: 7487, tv_nsec: 504422468 }, requested_resume: None })
[2020-02-07 15:55] [INFO] glutin event: UserEvent(Title("[haarp@hellbringer]:~ 15:55:07"))
[2020-02-07 15:55] [INFO] glutin event: UserEvent(Wakeup)
[2020-02-07 15:55] [INFO] glutin event: UserEvent(Wakeup)
[2020-02-07 15:55] [INFO] glutin event: MainEventsCleared
[2020-02-07 15:55] [INFO] glutin event: RedrawEventsCleared
REQUESTING REDRAW DUE TO VISUAL BELL
VISUAL BELL INTENSITY: 0.7966748706208749 -> true
[2020-02-07 15:55] [INFO] glutin event: NewEvents(WaitCancelled { start: Instant { tv_sec: 7487, tv_nsec: 521909413 }, requested_resume: None })
[2020-02-07 15:55] [INFO] glutin event: MainEventsCleared
[2020-02-07 15:55] [INFO] glutin event: RedrawEventsCleared
REQUESTING REDRAW DUE TO VISUAL BELL
VISUAL BELL INTENSITY: 0.6921817976480051 -> true
```

stuck.

Pressing a key (while Alacritty doesn't have focus) unstucks it:

```
[2020-02-07 15:55] [INFO] glutin event: NewEvents(WaitCancelled { start: Instant { tv_sec: 7487, tv_nsec: 537990954 }, requested_resume: None })
[2020-02-07 15:55] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(17))), event: Key(KeyboardInput { scancode: 28, state: Pressed, virtual_keycode: Some(Return), modifiers: (empty) }) }
[2020-02-07 15:55] [INFO] glutin event: MainEventsCleared
[2020-02-07 15:55] [INFO] glutin event: RedrawEventsCleared
REQUESTING REDRAW DUE TO VISUAL BELL

VISUAL BELL INTENSITY: 0 -> false
[2020-02-07 15:55] [INFO] glutin event: NewEvents(WaitCancelled { start: Instant { tv_sec: 7492, tv_nsec: 514833143 }, requested_resume: None })
[2020-02-07 15:55] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(17))), event: Key(KeyboardInput { scancode: 28, state: Released, virtual_keycode: Some(Return), modifiers: (empty) }) }
[2020-02-07 15:55] [INFO] glutin event: MainEventsCleared
[2020-02-07 15:55] [INFO] glutin event: RedrawEventsCleared
VISUAL BELL COMPLETED
VISUAL BELL INTENSITY: 0 -> false
[2020-02-07 15:55] [INFO] glutin event: NewEvents(WaitCancelled { start: Instant { tv_sec: 7492, tv_nsec: 554714043 }, requested_resume: None })
[2020-02-07 15:55] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(17))), event: Key(KeyboardInput { scancode: 29, state: Pressed, virtual_keycode: Some(LControl), modifiers: (empty) }) }
[2020-02-07 15:55] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(17))), event: ModifiersChanged(CTRL) }
[2020-02-07 15:55] [INFO] glutin event: MainEventsCleared
[2020-02-07 15:55] [INFO] glutin event: RedrawEventsCleared
[2020-02-07 15:55] [INFO] glutin event: NewEvents(WaitCancelled { start: Instant { tv_sec: 7495, tv_nsec: 826802856 }, requested_resume: None })
[2020-02-07 15:55] [INFO] glutin event: DeviceEvent { device_id: DeviceId(X(DeviceId(17))), event: Key(KeyboardInput { scancode: 46, state: Pressed, virtual_keycode: Some(C), modifiers: CTRL }) }
[2020-02-07 15:55] [INFO] glutin event: MainEventsCleared
[2020-02-07 15:55] [INFO] glutin event: RedrawEventsCleared
^C
```

**Comment by chrisduerr** [comment]
Reported upstream in https://github.com/rust-windowing/winit/issues/1442.

If someone is able to reproduce this on a system other than X11 - Linux, please let me know.

**Comment by kchibisov** [comment]
sorry for pinging, @dannydulai since you were the one who reproduced the issue on macOS, could you verify that you can get similar log to @haarp 's one with the patch and instructions from https://github.com/alacritty/alacritty/issues/3083#issuecomment-581208704? It'll be nice to also include macOS in upstream bug report.

**Comment by chrisduerr** [comment]
I'd really appreciate it if someone could try this patch too:

```diff
diff --git a/alacritty/src/event.rs b/alacritty/src/event.rs
index cd4f82a..ae7b46a 100644
--- a/alacritty/src/event.rs
+++ b/alacritty/src/event.rs
@@ -437,7 +437,7 @@ impl<N: Notify + OnResize> Processor<N> {
 
                 // Request immediate re-draw if visual bell animation is not finished yet
                 if !terminal.visual_bell.completed() {
-                    event_queue.push(GlutinEvent::UserEvent(Event::Wakeup));
+                    self.display.window.request_redraw();
                 }
 
                 // Redraw screen
diff --git a/alacritty/src/window.rs b/alacritty/src/window.rs
index 71327ba..5e5e548 100644
--- a/alacritty/src/window.rs
+++ b/alacritty/src/window.rs
@@ -131,6 +131,10 @@ pub struct Window {
 }
 
 impl Window {
+    pub fn request_redraw(&mut self) {
+        self.window().request_redraw();
+    }
+
     /// Create a new window
     ///
     /// This creates a window and fully initializes a window.
```

**Comment by haarp** [comment]
> I'd really appreciate it if someone could try this patch too, regardless of platform:

Nope, bell still stuck.

**Comment by kchibisov** [comment]
@haarp what `alacritty --version` was during test?

**Comment by haarp** [comment]
> @haarp what `alacritty --version` was during test?

0.4.1

**Comment by kchibisov** [comment]
Could you apply those patches from @chrisduerr to the latest master and retest? The initial one https://github.com/alacritty/alacritty/issues/3083#issuecomment-581208704 is more important here to test.

**Comment by chrisduerr** [comment]
Just to clarify, was it `0.4.1` during the initial tests, or did you actually apply the patches to `0.4.1` instead of master? If you applied them to master, then everything should be fine and upstream has all the infos necessary.

**Comment by haarp** [comment]
I have just retested current master (0.4.2-dev (5cd13f8)). The behavior with and without the `request_redraw` patch is unchanged from 0.4.1.

**Comment by meshulam** [comment]
I encountered this issue on MacOS with 0.4.1, I have a Macbook pro with an Intel GPU. I applied both patches above to the latest master and could still reproduce the issue. 

```
❯ uname -a
Darwin meshbook 18.7.0 Darwin Kernel Version 18.7.0: Thu Jan 23 06:52:12 PST 2020; root:xnu-4903.278.25~1/RELEASE_X86_64 x86_64
❯ cat bell.yml
visual_bell:
  animation: Linear
  duration: 200
  color: '0xff0000'
❯ target/release/osx/Alacritty.app/Contents/MacOS/alacritty  --config-file ./bell.yml --print-events
```

Pressing tab to trigger the bell:

```
[2020-03-12 17:27:52.941409000] [INFO] glutin event: NewEvents(WaitCancelled { start: Instant { t: 129899556312271 }, requested_resume: None })
[2020-03-12 17:27:52.953881000] [INFO] glutin event: WindowEvent { window_id: WindowId(Id(140424459067392)), event: KeyboardInput { device_id: DeviceId(DeviceId), input: KeyboardInput { scancode: 48, state: Pressed, virtual_keycode: Some(Tab), modifiers: (empty) }, is_synthetic: false } }
[2020-03-12 17:27:52.953969000] [INFO] glutin event: WindowEvent { window_id: WindowId(Id(140424459067392)), event: ReceivedCharacter('\t') }
[2020-03-12 17:27:52.953992000] [INFO] glutin event: MainEventsCleared
[2020-03-12 17:27:52.954007000] [INFO] glutin event: RedrawEventsCleared
VISUAL BELL COMPLETED
VISUAL BELL INTENSITY: 0 -> false
[2020-03-12 17:27:52.957447000] [INFO] glutin event: NewEvents(WaitCancelled { start: Instant { t: 129900548173845 }, requested_resume: None })
[2020-03-12 17:27:52.957546000] [INFO] glutin event: UserEvent(MouseCursorDirty)
[2020-03-12 17:27:52.957569000] [INFO] glutin event: UserEvent(Urgent)
[2020-03-12 17:27:52.957585000] [INFO] glutin event: UserEvent(Wakeup)
[2020-03-12 17:27:52.957602000] [INFO] glutin event: MainEventsCleared
[2020-03-12 17:27:52.957616000] [INFO] glutin event: RedrawEventsCleared
REQUESTING REDRAW DUE TO VISUAL BELL
VISUAL BELL INTENSITY: 0.992648445 -> true
[2020-03-12 17:27:53.014499000] [INFO] glutin event: NewEvents(WaitCancelled { start: Instant { t: 129900569044719 }, requested_resume: None })
[2020-03-12 17:27:53.014922000] [INFO] glutin event: WindowEvent { window_id: WindowId(Id(140424459067392)), event: KeyboardInput { device_id: DeviceId(DeviceId), input: KeyboardInput { scancode: 48, state: Released, virtual_keycode: Some(Tab), modifiers: (empty) }, is_synthetic: false } }
[2020-03-12 17:27:53.014989000] [INFO] glutin event: MainEventsCleared
[2020-03-12 17:27:53.015018000] [INFO] glutin event: RedrawRequested(WindowId(Id(140424459067392)))
[2020-03-12 17:27:53.015043000] [INFO] glutin event: RedrawEventsCleared
REQUESTING REDRAW DUE TO VISUAL BELL
VISUAL BELL INTENSITY: 0.70536239 -> true
[2020-03-12 17:27:53.484883000] [INFO] glutin event: NewEvents(WaitCancelled { start: Instant { t: 129900606727928 }, requested_resume: None })
[2020-03-12 17:27:53.485097000] [INFO] glutin event: MainEventsCleared
[2020-03-12 17:27:53.485163000] [INFO] glutin event: RedrawRequested(WindowId(Id(140424459067392)))
[2020-03-12 17:27:53.485203000] [INFO] glutin event: RedrawEventsCleared
REQUESTING REDRAW DUE TO VISUAL BELL
VISUAL BELL INTENSITY: 0 -> false
[2020-03-12 17:27:53.487718000] [INFO] glutin event: NewEvents(WaitCancelled { start: Instant { t: 129901078428005 }, requested_resume: None })
[2020-03-12 17:27:53.487836000] [INFO] glutin event: MainEventsCleared
[2020-03-12 17:27:53.487874000] [INFO] glutin event: RedrawRequested(WindowId(Id(140424459067392)))
[2020-03-12 17:27:53.487902000] [INFO] glutin event: RedrawEventsCleared
VISUAL BELL COMPLETED
VISUAL BELL INTENSITY: 0 -> false
```

**Comment by chrisduerr** [comment]
Thanks, I didn't really expect that to work but it was worth a shot. Tough to really pin it down without being able to reproduce it, but I believe this might be related to #824.

**Comment by chrisduerr** [comment]
I'd be curious if anyone was ever able to reproduce this with Wayland. It seems like this issue is exclusive to X11?

**Comment by rmsc** [comment]
This seems to be fixed in latest git (9028fb451a967d69a9e258a083ba64b052a9a5dd).
